File: Python/gaussSimple.py
import ast
import json
import numpy as np
from utils import regresiveSustitution
from utils import rowOps
from utils import getMultipliers
from utils import swapRows
from utils import isSquared

def gaussSimple(A, b):
    A = ast.literal_eval(A)
    b = ast.literal_eval(b)
    res = {}
    pivots = []
    # Convert into numpys arr
    A = np.array(A).astype(float)
    b = np.array(b).astype(float)
    # Appends last column to A matrix
    A = np.concatenate([A, b.reshape((A.shape[0], 1))], axis=1)
    # Validates if matrix is squared
    if not isSquared(np.delete(A, -1, axis=1)):
        res["status"] =  'Not square + 1 col matrix!'
        return res
    # Determines if det is 0
    if (np.linalg.det(np.delete(A, -1, axis=1)) == 0):
        res["status"] =  'Determinant is 0'
        return res
    times = A[:, 0].size - 1
    indexes = np.arange(0, times+1)
    for nCol in range(0, times):
        # Validates if A[i][i] is 0 and swap rows to first row in submatrix with col val != 0
        if (A[nCol][nCol] == 0):
            partCol = A[nCol:, nCol]
            nInd = np.argmax(partCol > 0)
            A, indexes = swapRows(A, nCol, nInd, indexes)
        # Calculates the multipliers
        multipliers = getMultipliers(A, nCol)
        # Validates if any multiplier is different to zero
        if (not np.count_nonzero(multipliers) == 0):
            A = rowOps(A, nCol, multipliers)
            pivots.append({"step" : nCol, "matrix": json.dumps(A.tolist())})

    values = regresiveSustitution(A, times, indexes)
    res["pivots"] = pivots
    res["values"] = values
    return res

File: C++/gaussSimple.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <sstream>
#include "utils.h"


using namespace std;

vector<vector<vector<double>>> gaussSimple(vector<vector<double>> matrix){
  vector<vector<vector<double>>> pivots; 
  vector<double> indexes; 
  pivots.push_back(matrix);
  if(!isSquared(matrix)){
    cout << "Not square + 1 col matrix" << endl;
    return pivots;
  }
  if(calcDeterminant(matrix) == 0){
    cout << "Det 0!" << endl;
    return pivots;
  }
  int times = matrix.size()-1;
  for(int i = 0; i <= times; i++ ){
    indexes.push_back(i);
  }

  for(int nCol = 0; nCol < times; nCol++){
      if(matrix[nCol][nCol] == 0.0){
        int index;
        for(int i = nCol; i < matrix.size(); i++){
          if(matrix[i][nCol] != 0){
            index = i;
            break;
          }
        }
        matrix[nCol].swap(matrix[index]);
        //Needs validation ------------------------------------------------
        int temInd = indexes[nCol];
        indexes[nCol] = indexes[index];
        indexes[index] = temInd;
        vector<vector<double>> temp =matrix; 
        pivots.push_back(temp);
      }
      vector<double> multipliers = getMultipliers(matrix, nCol);
      bool checkZero = true;
      for(int i = 0; i < multipliers.size(); i++){
        if(multipliers[i] != 0){
          checkZero = false;
        }
      }
      if(!checkZero){
        matrix = rowOps(matrix,nCol,multipliers);
        pivots.push_back(matrix);
      }
  }
  matrix = regressiveSust(matrix,times,indexes);
  pivots.push_back(matrix);
  return pivots;
}

int main (){
  /*string filename;
  cout << "Enter filename with matrix separated by spaces" <<endl; 
  cin >> filename;*/
  vector<vector<double>> matrix = parseMatrix("input.txt");
  vector<vector<vector<double>>> result = gaussSimple(matrix);
  writeTable(result);
}

